Implement OATH-HOTP two-factor authentication

Hylian opened the pull request:
## Description

First pass at an OATH-HOTP module to send OTP codes via keyboard. HOTP
is counter-based, with the counter being stored in EEPROM. The secret
and digit length are hard-coded into the binary. Pulls in HMAC-SHA1 and
SHA1 implementations, lightly modified to save code space. Tested working on an ErgoDox (EZ) running a Teensy w/ ATmega32u4.

## Types of Changes

- [x] Core
- [ ] Bugfix
- [x] New feature
- [ ] Enhancement/optimization
- [ ] Keyboard (addition or update)
- [ ] Keymap/layout/userspace (addition or update)
- [x] Documentation

## Checklist

- [hopefully!] My code follows the code style of this project.
- [maybe] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [ ] I have read the [**CONTRIBUTING** document](https://docs.qmk.fm/#/contributing).
- [ ] I have added tests to cover my changes.
- [x] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

drashna commented on quantum/hotp/hotp.c:
```diff
@@ -0,0 +1,197 @@
+
+#define HOTP_CONFIG_SECRET_BYTES   (sizeof(s_hotp_secret))
+#define HOTP_COUNTER_EEPROM_ADDR   (HOTP_EEPROM_ADDR)
+#define HOTP_COUNTER_SIZE_BYTES    (8)
```
This should probably be stored in `/tmk_core/common/eeconfig.h` rather than here. 

This is so to ensure that there are no collisions, like we've recently have had with VIA.

Hylian replied:
Gotcha, I'll move these over there. Is there a standard format to indicate the length of the data at an address?

drashna replied:
the type of uint used usually specifies that, but in this case, you'd want to specify a couple of blocks.  You can add a comment to where the end is, if you'd like (eg, the next available block)

drashna commented:
And yes, if you could add documentation, that would be awesome.

Hylian commented:
> And yes, if you could add documentation, that would be awesome.

Will do!

Hylian commented:
Changes:
- Moved EEPROM address to `eeconfig.h`
- Removed `hotp_config.h` and moved user-facing configuration to `config.h`
- Added documentation
- EEPROM init now resets HOTP counter to 0

drashna commented on tmk_core/common/eeconfig.h:
```diff
@@ -48,6 +48,7 @@ along with this program.  If not, see <http://www.gnu.org/licenses/>.
 #define EECONFIG_RGB_MATRIX_SPEED (uint8_t *)32
 // TODO: Combine these into a single word and single block of EEPROM
 #define EECONFIG_KEYMAP_UPPER_BYTE (uint8_t *)33
+#define EECONFIG_HOTP (uint8_t *)34 // 8 bytes
```
For ... sanity's stake, this probably should be set to something like: 
```suggestion
#define EECONFIG_HOTP (uint32_t *)34 // Uses 8 bytes
#define EECONFIG_HOTP2 (uint32_t *)38 // Additional buffer
```

And this way, you could use:
```c
eeprom_update_dword(EECONFIG_HOTP, 0);
eeprom_update_dword(EECONFIG_HOTP2, 0);
```

Hylian replied:
Heh, yeah, that's clearer. I'll switch to that.

Hylian commented:
Changes:
 * Updated address list to use two `uint32_t *`'s
 * Updated eeconfig init to use new defines

drashna commented on docs/feature_hotp.md:
```diff
@@ -0,0 +1,31 @@
+Add your HOTP configuration to `config.h`:
+
+* `HOTP_OUTPUT_DIGITS` — Number of output digits. Typically between 6 and 8.
+* `HOTP_SECRET` — Your secret. This _must_ be formatted in the form of a hex array: 
```
It might be helpful here to have something that will convert it from a string (or however you get it) to a hex array.

Hylian replied:
It's typically provisioned as a base32 encoded string. We could maybe pull in an additional dependency to a b32 library, such as this one: https://github.com/mjg59/tpmtotp/blob/master/base32.c

drashna commented on quantum/hotp/hotp.c:
```diff
@@ -0,0 +1,207 @@
+
+#ifndef EECONFIG_HOTP
+    #error EECONFIG_HOTP not defined in eeconfig.h
+#endif
```
Shouldn't need this block anymore

hugovk commented on docs/feature_hotp.md:
````diff
@@ -0,0 +1,31 @@
+{0x00, 0x01, ..., 0xFF}
+```
+
+!> Make sure you don't commit your secret to git, or everyone will be able to see it! One way to manage this might be to define `HOTP_SECRET` in a separate header file that's included in your `config.h`.
````
```suggestion
!> Make sure you don't commit your secret to Git, or everyone will be able to see it! One way to manage this is to define `HOTP_SECRET` in a separate header file that's included in your `config.h`.
```

dylangerdaly commented:
Any update on this? Even a simple HMAC Function would be great for LUKs Unlocking.

Hylian commented:
Hey @dylangerdaly, sorry, I sort of lost interest in this when I realized my workplace's MFA actually only supported Yubikey's HOTP, which is different from OATH-HOTP for some reason. Made it a bit more difficult to integration test as well without a real world example. OATH-HOTP doesn't really seem to be used much in the real world, so I figured I'd just drop it.

How can this be used with LUKS? I use LUKS as well, so that could be interesting. Keyboards aren't secure at all, obviously, but if it could be used as a second factor, that could be neat. There's always the dumb option of just having a preprogrammed string that gets emitted on a combo, and appending that to a password that you know, too.

dylangerdaly commented:
Hey Edward,

No worries, for LUKs I was thinking of having a secret salt baked into the firmware along with a passphrase to present an HMAC back to the Host, however you're right, Keyboards aren't secure.

It would be great if there was a Secure Element, I don't think we'll ever see that, I'd better stick to Yubikey etc.

Thanks you for you contribution in any case 😀

Sent from ProtonMail mobile

-------- Original Message --------
On Jun 16, 2020, 4:12 AM, Edward Shin wrote:

> Hey [@dylangerdaly](https://github.com/dylangerdaly), sorry, I sort of lost interest in this when I realized my workplace's MFA actually only supported Yubikey's HOTP, which is different from OATH-HOTP for some reason. Made it a bit more difficult to integration test as well without a real world example. OATH-HOTP doesn't really seem to be used much in the real world, so I figured I'd just drop it.
>
> How can this be used with LUKS? I use LUKS as well, so that could be interesting. Keyboards aren't secure at all, obviously, but if it could be used as a second factor, that could be neat. There's always the dumb option of just having a preprogrammed string that gets emitted on a combo, and appending that to a password that you know, too.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, [view it on GitHub](https://github.com/qmk/qmk_firmware/pull/7066#issuecomment-644288827), or [unsubscribe](https://github.com/notifications/unsubscribe-auth/AB4ZXETT3LEMMZFCSJBHSNTRWZQARANCNFSM4JCAD4MQ).

drashna commented:
Would it be okay if we closed this PR, then?

Hylian commented:
Yep, that would be fine, thanks!

On Wed, Jun 17, 2020 at 8:58 PM Drashna Jaelre <notifications@github.com>
wrote:

> Would it be okay if we closed this PR, then?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/qmk/qmk_firmware/pull/7066#issuecomment-645705207>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AACF4CVRMV7WTO4X44IQ7YLRXFRD5ANCNFSM4JCAD4MQ>
> .
>

zvecr commented:
Closing per recent discussion.
